app/services/sidebar_site_settings_backfiller.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+# A service class that backfills the changes to the default sidebar categories and tags site settings.
+#
+# When a category/tag is removed from the site settings, the `SidebarSectionLink` records associated with the category/tag
+# are deleted.
+#
+# When a category/tag is added to the site settings, a `SidebarSectionLink` record for the associated category/tag are
+# created for all users that do not already have a `SidebarSectionLink` record for the category/tag.
+class SidebarSiteSettingsBackfiller
+  def initialize(setting_name, previous_value:, new_value:)
+    @setting_name = setting_name
+
+    @linkable_klass, previous_ids, new_ids =
+      case setting_name
+      when "default_sidebar_categories"
+        [
+          Category,
+          previous_value.split("|"),
+          new_value.split("|")
+        ]
+      when "default_sidebar_tags"
+        klass = Tag
+
+        [
+          klass,
+          klass.where(name: previous_value.split("|")).pluck(:id),
+          klass.where(name: new_value.split("|")).pluck(:id)
+        ]
+      end
+
+    @added_ids = new_ids - previous_ids
+    @removed_ids = previous_ids - new_ids
+  end
+
+  def backfill!
+    DistributedMutex.synchronize("backfill_sidebar_site_settings_#{@setting_name}") do
+      SidebarSectionLink.where(linkable_type: @linkable_klass.to_s, linkable_id: @removed_ids).delete_all
+
+      User.real.where(staged: false).select(:id).find_in_batches do |users|
```

Comment:
I am not sure about this. Let's talk about scenario when default tag is "a".

User edit preferences and set that they prefer tag "b"

Finally, admin change default sidebar tag setting to "c"

In this case, we will override user choice. Shouldn't we have a guard condition and pick only users who had exactly same previous setting before update from admin?

Reply:
In the scenario above, the user will end up with tag "b" and tag "c" if the admin elects to apply the changes from tag "a" to tag "c" historically. Anyone with tag "c" configured will have the setting removed.

Reply:
Ah, I understand it now, thank you 👍